In [223]:
import numpy as np 
import pandas as pd

In [224]:
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

In [225]:
df= pd.read_csv('flats.csv')

In [226]:
df.sample(10)

property_name  \
2526   3 BHK Flat in Sector 61 Gurgaon   
1968  3 BHK Flat in Sector 102 Gurgaon   
2629   3 BHK Flat in Sector 92 Gurgaon   
2024  3 BHK Flat in Sector 102 Gurgaon   
2137   2 BHK Flat in Sector 79 Gurgaon   
2571   3 BHK Flat in Sector 86 Gurgaon   
1786   3 BHK Flat in Sector 86 Gurgaon   
3013   5 BHK Flat in Sector 48 Gurgaon   
1297   3 BHK Flat in Sector 81 Gurgaon   
1901  3 BHK Flat in Sector 109 Gurgaon   

                                                                                                                                                    link  \
2526                    https://www.99acres.com/3-bhk-bedroom-apartment-flat-for-sale-in-smart-world-orchard-sector-61-gurgaon-1630-sq-ft-spid-T69058716   
1968                  https://www.99acres.com/3-bhk-bedroom-apartment-flat-for-sale-in-emaar-gurgaon-greens-sector-102-gurgaon-1650-sq-ft-spid-I70126602   
2629                                        https://www.99acres.com/3-bhk-bedroom-apartment-flat-for-sale-in-sector-92-gurgaon-1600-sq-ft-spid-S69087390   
2024  https://www.99acres.com/3-bhk-bedroom-apartment-flat-for-sale-in-shapoorji-pallonji-joyville-gurugram-sector-102-gurgaon-1827-sq-ft-spid-S70038400   
2137                                        https://www.99acres.com/2-bhk-bedroom-apartment-flat-for-sale-in-sector-79-gurgaon-1420-sq-ft-spid-D69857690   
2571                       https://www.99acres.com/3-bhk-bedroom-apartment-flat-for-sale-in-ansal-heights-86-sector-86-gurgaon-1690-sq-ft-spid-L68958088   
1786             https://www.99acres.com/3-bhk-bedroom-apartment-flat-for-sale-in-dlf-new-town-heights-2-sector-86-gurgaon-1930-sq-ft-r12-spid-A41264401   
3013               https://www.99acres.com/5-bhk-bedroom-apartment-flat-for-sale-in-parsvnath-green-ville-sector-48-gurgaon-3905-sq-ft-r7-spid-J17123294   
1297                         https://www.99acres.com/3-bhk-bedroom-apartment-flat-for-sale-in-dlf-the-ultima-sector-81-gurgaon-2132-sq-ft-spid-D69512748   
1901                         https://www.99acres.com/3-bhk-bedroom-apartment-flat-for-sale-in-sbtl-caladium-sector-109-gurgaon-1880-sq-ft-spid-H69848302   

                                        society       price             area  \
2526                        Smart World Orchard  2.41 Crore  ₹ 14,785/sq.ft.   
1968                  Emaar Gurgaon Greens4.1 ★   1.6 Crore   ₹ 9,696/sq.ft.   
2629                                  Apartment      75 Lac   ₹ 4,687/sq.ft.   
2024  Shapoorji Pallonji Joyville Gurugram4.0 ★   2.5 Crore  ₹ 14,000/sq.ft.   
2137                             M3m Golf Hills  1.45 Crore  ₹ 10,211/sq.ft.   
2571                           Ansal Heights 86      90 Lac   ₹ 5,325/sq.ft.   
1786                DLF New Town Heights 24.0 ★  1.26 Crore   ₹ 6,113/sq.ft.   
3013                 Parsvnath Green Ville3.9 ★   3.3 Crore   ₹ 9,984/sq.ft.   
1297                        DLF The Ultima4.0 ★  2.45 Crore  ₹ 11,491/sq.ft.   
1901                              SBTL Caladium   1.3 Crore   ₹ 6,914/sq.ft.   

                                                                                                                areaWithType  \
2526                                               Built Up area: 1630 (151.43 sq.m.)Carpet area: 1625 sq.ft. (150.97 sq.m.)   
1968                                            Super Built up area 1650(153.29 sq.m.)Carpet area: 1022.58 sq.ft. (95 sq.m.)   
2629                                                                                        Carpet area: 1600 (148.64 sq.m.)   
2024                                                                                  Super Built up area 1827(169.73 sq.m.)   
2137                                                                                        Carpet area: 1420 (131.92 sq.m.)   
2571                                                                                      Built Up area: 1690 (157.01 sq.m.)   
1786                                                                

In [227]:
# Checking Shape of data
df.shape

(3017, 20)

In [228]:
# Checking basic information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3017 entries, 0 to 3016
Data columns (total 20 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   property_name    3017 non-null   object
 1   link             3017 non-null   object
 2   society          3016 non-null   object
 3   price            3007 non-null   object
 4   area             3004 non-null   object
 5   areaWithType     3008 non-null   object
 6   bedRoom          3008 non-null   object
 7   bathroom         3008 non-null   object
 8   balcony          3008 non-null   object
 9   additionalRoom   1694 non-null   object
 10  address          3002 non-null   object
 11  floorNum         3006 non-null   object
 12  facing           2127 non-null   object
 13  agePossession    3007 non-null   object
 14  nearbyLocations  2913 non-null   object
 15  description      3008 non-null   object
 16  furnishDetails   2203 non-null   object
 17  features         2594 non-null   

 - All column is "String Data Type" which need to change some of them.
- A lot of column have missing values 

In [229]:
# check for duplicates
df.duplicated().sum()

0

In [230]:
# check for missing values
df.isnull().sum()

property_name         0
link                  0
society               1
price                10
area                 13
areaWithType          9
bedRoom               9
bathroom              9
balcony               9
additionalRoom     1323
address              15
floorNum             11
facing              890
agePossession        10
nearbyLocations     104
description           9
furnishDetails      814
features            423
rating              341
property_id           9
dtype: int64

- In few columns like "additionalRoom" -1323,"facing" -890, "furnishDetails" - 814, "features" - 413, "rating" - 341 
- Missing values and other columns also have missing value under 100 rows except 2 columns. 
- Property_name and link columns have zero missing values.


In [231]:
# drop columns - link, property_id 
df.drop(columns= ['link', 'property_id'], inplace = True)

- removed the columns which I am not going to use for further analysis.
- "link" removing this column as it is useful if I need to upload images or any further information needed from website, as I am not creating website for property display , I am creating analytics website where I display depth analysis through grpahs.
- "property_id" also removing this column as it helps to find duplicate value as no duplicate in dataset so can remove this, or if I need to merge two dataset then I can use this column which is not applicable as of now as I have data in same file.


In [232]:
df.head()

property_name                 society       price  \
0     2 BHK Flat in Krishna Colony  maa bhagwati residency      45 Lac   
1        2 BHK Flat in Ashok Vihar            Apna Enclave      50 Lac   
2              2 BHK Flat in Sohna  Tulsiani Easy in Homes      40 Lac   
3  2 BHK Flat in Sector 61 Gurgaon     Smart World Orchard  1.47 Crore   
4  2 BHK Flat in Sector 92 Gurgaon        Parkwood Westend      70 Lac   

              area                            areaWithType     bedRoom  \
0   ₹ 5,000/sq.ft.          Carpet area: 900 (83.61 sq.m.)  2 Bedrooms   
1   ₹ 7,692/sq.ft.          Carpet area: 650 (60.39 sq.m.)  2 Bedrooms   
2   ₹ 6,722/sq.ft.          Carpet area: 595 (55.28 sq.m.)  2 Bedrooms   
3  ₹ 12,250/sq.ft.        Carpet area: 1200 (111.48 sq.m.)  2 Bedrooms   
4   ₹ 5,204/sq.ft.  Super Built up area 1345(124.95 sq.m.)  2 Bedrooms   

      bathroom      balcony additionalRoom  \
0  2 Bathrooms    1 Balcony            NaN   
1  2 Bathrooms    1 Balcony            NaN   
2  2 Bathrooms  3 Balconies            NaN   
3  2 Bathrooms  2 Balconies     Study Room   
4  2 Bathrooms  3 Balconies     Study Room   

                               address             floorNum facing  \
0     Krishna Colony, Gurgaon, Haryana    4th   of 4 Floors   West   
1   46b, Ashok Vihar, Gurgaon, Haryana    1st   of 3 Floors   West   
2              Sohna, Gurgaon, Haryana  12nd   of 14 Floors    NaN   
3  Sector 61 Gurgaon, Gurgaon, Haryana    2nd   of 4 Floors    NaN   
4  Sector 92 Gurgaon, Gurgaon, Haryana    5th   of 8 Floors    NaN   

        agePossession  \
0     1 to 5 Year Old   
1        10+ Year Old   
2     0 to 1 Year Old   
3            Dec 2023   
4  Under Construction   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                nearbyLocations  \
0  ['Chintapurni Mandir', 'State bank ATM', 'Pearl Dental Clinic', 'Taneja Hospital', 'Dr. Ashok Jain', 'Dr. Madan Clinic', 'Geeta Nursing Home Gurgaon', 'Chiranjiv Hospital', 'Dev Man Kathuria Clinic', 'Aryan Hospital', 'Kr Dental Hub', 'Swastik Maternity and Medical Centre', 'Satyam Hospital Gurgaon', 'Ravi Clinic and Health Care Centre', 'D.R. Rajnis Gupta Clinic', 'Esic Hospital Gurugram', 'Gurgaon Eye Centre', 'Bindal Clinic', 'Shiv Mahima Patient Care Bureau', 'Shree Krishna Hospital Gurgaon', 'M.S Hospital', 'My Care Clinic', 'Ankur Clinic and Maternity Home', 'Triveni Hospital Gurgaon', 'Shri Gobind Hospital', 'Navjeevan Hospital and Maternity Centre', 'Lal Superspeciality Hospital', 'Sparsh Hospital Gurgaon', 'Prateek Nursing Home And Polyclinic', 'Jain Sant Phool Chand Ji Charitable Hospital', 'Sarvodya Hospital', 'Bhatnagar Maternity and Nursing Home', 'Tirath Ram Hospitals Pvt L

In [233]:
# rename columns
df.rename(columns={'area':'price_per_sqft'},inplace =True)
df.head()

property_name                 society       price  \
0     2 BHK Flat in Krishna Colony  maa bhagwati residency      45 Lac   
1        2 BHK Flat in Ashok Vihar            Apna Enclave      50 Lac   
2              2 BHK Flat in Sohna  Tulsiani Easy in Homes      40 Lac   
3  2 BHK Flat in Sector 61 Gurgaon     Smart World Orchard  1.47 Crore   
4  2 BHK Flat in Sector 92 Gurgaon        Parkwood Westend      70 Lac   

    price_per_sqft                            areaWithType     bedRoom  \
0   ₹ 5,000/sq.ft.          Carpet area: 900 (83.61 sq.m.)  2 Bedrooms   
1   ₹ 7,692/sq.ft.          Carpet area: 650 (60.39 sq.m.)  2 Bedrooms   
2   ₹ 6,722/sq.ft.          Carpet area: 595 (55.28 sq.m.)  2 Bedrooms   
3  ₹ 12,250/sq.ft.        Carpet area: 1200 (111.48 sq.m.)  2 Bedrooms   
4   ₹ 5,204/sq.ft.  Super Built up area 1345(124.95 sq.m.)  2 Bedrooms   

      bathroom      balcony additionalRoom  \
0  2 Bathrooms    1 Balcony            NaN   
1  2 Bathrooms    1 Balcony            NaN   
2  2 Bathrooms  3 Balconies            NaN   
3  2 Bathrooms  2 Balconies     Study Room   
4  2 Bathrooms  3 Balconies     Study Room   

                               address             floorNum facing  \
0     Krishna Colony, Gurgaon, Haryana    4th   of 4 Floors   West   
1   46b, Ashok Vihar, Gurgaon, Haryana    1st   of 3 Floors   West   
2              Sohna, Gurgaon, Haryana  12nd   of 14 Floors    NaN   
3  Sector 61 Gurgaon, Gurgaon, Haryana    2nd   of 4 Floors    NaN   
4  Sector 92 Gurgaon, Gurgaon, Haryana    5th   of 8 Floors    NaN   

        agePossession  \
0     1 to 5 Year Old   
1        10+ Year Old   
2     0 to 1 Year Old   
3            Dec 2023   
4  Under Construction   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                nearbyLocations  \
0  ['Chintapurni Mandir', 'State bank ATM', 'Pearl Dental Clinic', 'Taneja Hospital', 'Dr. Ashok Jain', 'Dr. Madan Clinic', 'Geeta Nursing Home Gurgaon', 'Chiranjiv Hospital', 'Dev Man Kathuria Clinic', 'Aryan Hospital', 'Kr Dental Hub', 'Swastik Maternity and Medical Centre', 'Satyam Hospital Gurgaon', 'Ravi Clinic and Health Care Centre', 'D.R. Rajnis Gupta Clinic', 'Esic Hospital Gurugram', 'Gurgaon Eye Centre', 'Bindal Clinic', 'Shiv Mahima Patient Care Bureau', 'Shree Krishna Hospital Gurgaon', 'M.S Hospital', 'My Care Clinic', 'Ankur Clinic and Maternity Home', 'Triveni Hospital Gurgaon', 'Shri Gobind Hospital', 'Navjeevan Hospital and Maternity Centre', 'Lal Superspeciality Hospital', 'Sparsh Hospital Gurgaon', 'Prateek Nursing Home And Polyclinic', 'Jain Sant Phool Chand Ji Charitable Hospital', 'Sarvodya Hospital', 'Bhatnagar Maternity and Nursing Home', 'Tirath Ram Hospitals Pvt L

- Renamed columns "area" into "price_per_sqft" as it is not right name.

## Univariate Analysis

## Society

In [234]:
df['society'].value_counts()

society
SS The Leaf3.8 ★                                     73
Tulip Violet4.3 ★                                    40
Shapoorji Pallonji Joyville Gurugram4.0 ★            39
Signature Global Park4.0 ★                           36
Shree Vardhman Victoria3.8 ★                         35
Tulip Violet4.2 ★                                    33
Emaar MGF Emerald Floors Premier3.8 ★                32
Smart World Orchard                                  32
Smart World Gems                                     32
Paras Dews                                           31
DLF The Ultima4.0 ★                                  31
DLF Regal Gardens3.9 ★                               30
M3M Woodshire4.0 ★                                   29
Shree Vardhman Flora3.8 ★                            29
La Vida by Tata Housing                              28
Signature Global Solera3.7 ★                         28
Godrej Nature Plus                                   27
Emaar Gurgaon Greens4.1 ★               

### Observation:-

- In this data we have n number of society data and some society name written two times like one in the capital and other small.
- Some society name written along with ratings which create extra name.Ex- BPTP Park Generations3.7 *
- So for removing all these from society columns will do some preprocessing
- First of all check how many unique societies available.


In [235]:
df['society'].value_counts().shape

(638,)

- 638 is too much society ,so I need to remove this or correct the name.
- I used "regular expression" for this and if anywhere number or star mark "*" found remove them and convert all value in lower case .

In [236]:
import re

In [237]:
df['society'] = df['society'].apply(lambda name: re.sub(r'\d+(\.\d+)?\s?★', '', str(name)).strip()).str.lower()

In [238]:
df['society'].value_counts()

society
tulip violet                                     75
ss the leaf                                      74
shapoorji pallonji joyville gurugram             45
dlf new town heights                             38
signature global park                            36
shree vardhman victoria                          35
smart world orchard                              32
emaar mgf emerald floors premier                 32
smart world gems                                 32
paras dews                                       31
dlf the ultima                                   31
m3m woodshire                                    30
dlf regal gardens                                30
shree vardhman flora                             29
la vida by tata housing                          28
signature global solera                          28
godrej nature plus                               27
bptp terra                                       25
emaar gurgaon greens                             25
expe

In [239]:
df['society'].value_counts().shape

(604,)

In [240]:
df.head(1)

property_name                 society   price  \
0  2 BHK Flat in Krishna Colony  maa bhagwati residency  45 Lac   

   price_per_sqft                    areaWithType     bedRoom     bathroom  \
0  ₹ 5,000/sq.ft.  Carpet area: 900 (83.61 sq.m.)  2 Bedrooms  2 Bathrooms   

     balcony additionalRoom                           address  \
0  1 Balcony            NaN  Krishna Colony, Gurgaon, Haryana   

            floorNum facing    agePossession  \
0  4th   of 4 Floors   West  1 to 5 Year Old   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                nearbyLocations  \
0  ['Chintapurni Mandir', 'State bank ATM', 'Pearl Dental Clinic', 'Taneja Hospital', 'Dr. Ashok Jain', 'Dr. Madan Clinic', 'Geeta Nursing Home Gurgaon', 'Chiranjiv Hospital', 'Dev Man Kathuria Clinic', 'Aryan Hospital', 'Kr Dental Hub', 'Swastik Maternity and Medical Centre', 'Satyam Hospital Gurgaon', 'Ravi Clinic and Health Care Centre', 'D.R. Rajnis Gupta Clinic', 'Esic Hospital Gurugram', 'Gurgaon Eye Centre', 'Bindal Clinic', 'Shiv Mahima Patient Care Bureau', 'Shree Krishna Hospital Gurgaon', 'M.S Hospital', 'My Care Clinic', 'Ankur Clinic and Maternity Home', 'Triveni Hospital Gurgaon', 'Shri Gobind Hospital', 'Navjeevan Hospital and Maternity Centre', 'Lal Superspeciality Hospital', 'Sparsh Hospital Gurgaon', 'Prateek Nursing Home And Polyclinic', 'Jain Sant Phool Chand Ji Charitable Hospital', 'Sarvodya Hospital', 'Bhatnagar Maternity and Nursing Home', 'Tirath Ram Hospitals Pvt Ltd', 'Dr. Sarvejeet Singh', 'Centre For Sight Gurgaon New Railway Road', 'Dr. Ajay S. Gupta Clinic', 'Kidney Clinic', 'Mangalam Hospital and Heart Centre Gurgaon', 'Sethi Hospital Gurgaon', 'Shubham Hospital Gurgaon', 'Kharbanda Maternity and Nursing Home', 'Gupta Hospital Gurgaon', 'Dentecare - Multispeciality Dental Clinic', 'State bank of india', 'Hdfc bank', 'Kotak bank', 'Indian bank', 'Pizza Hut', 'St. Michaels Sr. Sec. School', 'Dronacharya Government College']   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  description  \
0  So with lift.Maa bhagwati residency is one of gurgaon's most sought after destination for apartments and this 2 bhk flat in krishna colony is your opportunity to be a part of this community. The flat occupies a carpet area of 900 sq.Ft. That consists of 2

# Price

In [241]:
df['price']

0                 45 Lac
1                 50 Lac
2                 40 Lac
3             1.47 Crore
4                 70 Lac
5                 41 Lac
6                2 Crore
7              1.8 Crore
8              1.1 Crore
9             4.75 Crore
10                96 Lac
11                29 Lac
12            1.35 Crore
13                95 Lac
14                70 Lac
15            3.95 Crore
16                90 Lac
17            1.05 Crore
18                   NaN
19             2.2 Crore
20            1.01 Crore
21            1.85 Crore
22                86 Lac
23            1.85 Crore
24            2.85 Crore
25      Price on Request
26                42 Lac
27            6.15 Crore
28            6.25 Crore
29                42 Lac
30            1.35 Crore
31             1.6 Crore
32            3.25 Crore
33                45 Lac
34                85 Lac
35                75 Lac
36                82 Lac
37             1.8 Crore
38             29.99 Lac
39                78 Lac


## Observations:-
- I observed two distinct formats for expressing prices: 'lakh' for amounts below one crore and 'cr' for amount equal to or exceeding one crore.
- To maintain consistency,I opted to represent all values using the 'cr' symbol as the common unit.
- I applied the same approcah to inspect the distribution of price categories across all value counts.

In [242]:
df['price'].value_counts()

price
1.25 Crore          79
1.1 Crore           61
1.4 Crore           60
1.5 Crore           59
1.2 Crore           59
90 Lac              58
1.3 Crore           57
95 Lac              53
2 Crore             51
1.75 Crore          47
1 Crore             46
1.6 Crore           43
1.35 Crore          41
1.55 Crore          40
1.9 Crore           40
75 Lac              38
1.65 Crore          38
1.8 Crore           37
1.7 Crore           37
80 Lac              36
2.2 Crore           34
50 Lac              33
1.15 Crore          33
1.45 Crore          31
85 Lac              31
1.05 Crore          30
2.5 Crore           29
60 Lac              29
40 Lac              29
2.1 Crore           26
45 Lac              26
65 Lac              25
35 Lac              23
2.35 Crore          23
1.85 Crore          23
3 Crore             22
70 Lac              21
55 Lac              20
2.25 Crore          20
3.5 Crore           19
2.4 Crore           18
2.3 Crore           18
30 Lac              17
2.65 

In [243]:
df[df['price']== 'Price on Request']

property_name                    society  \
25      5 BHK Flat in Sector 1 Gurgaon  the new people apartments   
1644   4 BHK Flat in Sector 50 Gurgaon                ss hibiscus   
1972   1 BHK Flat in Sector 31 Gurgaon            raheja atlantis   
2054   2 BHK Flat in Sector 79 Gurgaon         supertech araville   
2357  2 BHK Flat in Sector 37D Gurgaon            the millennia 2   
2522  2 BHK Flat in Sector 95A Gurgaon              the roselia 2   
2549      3 BHK Flat in Sector-2 Sohna         supertech hilltown   
2640  2 BHK Flat in Sector 95A Gurgaon              the roselia 2   
2959  3 BHK Flat in Sector 104 Gurgaon              godrej summit   
2972         4 BHK Flat in Gwal Pahari     krrish provence estate   
2973   4 BHK Flat in Sector 95 Gurgaon    sidhartha ncr one phase   

                 price price_per_sqft  \
25    Price on Request            NaN   
1644  Price on Request            NaN   
1972  Price on Request            NaN   
2054  Price on Request            NaN   
2357  Price on Request            NaN   
2522  Price on Request            NaN   
2549  Price on Request            NaN   
2640  Price on Request            NaN   
2959  Price on Request            NaN   
2972  Price on Request            NaN   
2973  Price on Request            NaN   

                                                                     areaWithType  \
25                                             Built Up area: 4500 (418.06 sq.m.)   
1644                                             Carpet area: 4800 (445.93 sq.m.)   
1972                                               Carpet area: 650 (60.39 sq.m.)   
2054                                           Built Up area: 1295 (120.31 sq.m.)   
2357                                               Carpet area: 552 (51.28 sq.m.)   
2522                                               Carpet area: 593 (55.09 sq.m.)   
2549                                           Built Up area: 1720 (159.79 sq.m.)   
2640                                               Carpet area: 593 (55.09 sq.m.)   
2959  Super Built up area 1647(153.01 sq.m.)Carpet area: 884 sq.ft. (82.13 sq.m.)   
2972                                           Built Up area: 5800 (538.84 sq.m.)   
2973                                       Super Built up area 2354(218.69 sq.m.)   

         bedRoom     bathroom       balcony  \
25    5 Bedrooms   1 Bathroom    No Balcony   
1644  4 Bedrooms  4 Bathrooms   3 Balconies   
1972   1 Bedroom   1 Bathroom     1 Balcony   
2054  2 Bedrooms  2 Bathrooms    No Balcony   
2357  2 Bedrooms   1 Bathroom     1 Balcony   
2522  2 Bedrooms  2 Bathrooms   2 Balconies   
2549  3 Bedrooms  3 Bathrooms   3 Balconies   
2640  2 Bedrooms  2 Bathrooms   2 Balconies   
2959  3 Bedrooms  3 Bathrooms   3 Balconies   
2972  4 Bedrooms  5 Bathrooms     1 Balcony   
2973  4 Bedrooms  4 Bathrooms  3+ Balconies   

                          additionalRoom  \
25                                   NaN   
1644                        Servant Room   
1972  Pooja Room,Study Room,Servant Room   
2054                                 NaN   
2357                                 NaN   
2522                                 NaN   
2549                                 NaN   
2640                                 NaN   
2959                                 NaN   
2972                                 NaN   
2973                                 NaN   

                                                             address  \
25                      Sector 1, Sector 1 Gurgaon, Gurgaon, Haryana   
1644                             Sector 50 Gurgaon, Gurgaon, Haryana   
1972                        Yes, Sector 31 Gurgaon, Gurgaon, Haryana   
2054                  Sector 79, Sector 79 Gurgaon, Gurgaon, Haryana   
2357                            Sector 37D Gurgaon, Gurgaon, Haryana   
2522  Sector 95a, Pataudi Road, Sector 95A Gurgaon, Gurgaon, Haryana   
2549        Sohna Sector 2,gurgaon, Sector-2 Sohna, Gurgaon, Haryana   
2640  Sector

- I examined approximately 10 flats, and in some cases, the price was specified as "price on request."  
- Consequently, I opted to eliminate these 10 rows from the dataset. The rationale behind this decision is related to the preparation of a linear regression model where the output is dependent on the price. Without a specified price, it becomes challenging to work with the data in the context of the regression model.


In [244]:
df = df[df['price'] != 'Price on Request']

In [245]:
df.head()

property_name                 society       price  \
0     2 BHK Flat in Krishna Colony  maa bhagwati residency      45 Lac   
1        2 BHK Flat in Ashok Vihar            apna enclave      50 Lac   
2              2 BHK Flat in Sohna  tulsiani easy in homes      40 Lac   
3  2 BHK Flat in Sector 61 Gurgaon     smart world orchard  1.47 Crore   
4  2 BHK Flat in Sector 92 Gurgaon        parkwood westend      70 Lac   

    price_per_sqft                            areaWithType     bedRoom  \
0   ₹ 5,000/sq.ft.          Carpet area: 900 (83.61 sq.m.)  2 Bedrooms   
1   ₹ 7,692/sq.ft.          Carpet area: 650 (60.39 sq.m.)  2 Bedrooms   
2   ₹ 6,722/sq.ft.          Carpet area: 595 (55.28 sq.m.)  2 Bedrooms   
3  ₹ 12,250/sq.ft.        Carpet area: 1200 (111.48 sq.m.)  2 Bedrooms   
4   ₹ 5,204/sq.ft.  Super Built up area 1345(124.95 sq.m.)  2 Bedrooms   

      bathroom      balcony additionalRoom  \
0  2 Bathrooms    1 Balcony            NaN   
1  2 Bathrooms    1 Balcony            NaN   
2  2 Bathrooms  3 Balconies            NaN   
3  2 Bathrooms  2 Balconies     Study Room   
4  2 Bathrooms  3 Balconies     Study Room   

                               address             floorNum facing  \
0     Krishna Colony, Gurgaon, Haryana    4th   of 4 Floors   West   
1   46b, Ashok Vihar, Gurgaon, Haryana    1st   of 3 Floors   West   
2              Sohna, Gurgaon, Haryana  12nd   of 14 Floors    NaN   
3  Sector 61 Gurgaon, Gurgaon, Haryana    2nd   of 4 Floors    NaN   
4  Sector 92 Gurgaon, Gurgaon, Haryana    5th   of 8 Floors    NaN   

        agePossession  \
0     1 to 5 Year Old   
1        10+ Year Old   
2     0 to 1 Year Old   
3            Dec 2023   
4  Under Construction   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                nearbyLocations  \
0  ['Chintapurni Mandir', 'State bank ATM', 'Pearl Dental Clinic', 'Taneja Hospital', 'Dr. Ashok Jain', 'Dr. Madan Clinic', 'Geeta Nursing Home Gurgaon', 'Chiranjiv Hospital', 'Dev Man Kathuria Clinic', 'Aryan Hospital', 'Kr Dental Hub', 'Swastik Maternity and Medical Centre', 'Satyam Hospital Gurgaon', 'Ravi Clinic and Health Care Centre', 'D.R. Rajnis Gupta Clinic', 'Esic Hospital Gurugram', 'Gurgaon Eye Centre', 'Bindal Clinic', 'Shiv Mahima Patient Care Bureau', 'Shree Krishna Hospital Gurgaon', 'M.S Hospital', 'My Care Clinic', 'Ankur Clinic and Maternity Home', 'Triveni Hospital Gurgaon', 'Shri Gobind Hospital', 'Navjeevan Hospital and Maternity Centre', 'Lal Superspeciality Hospital', 'Sparsh Hospital Gurgaon', 'Prateek Nursing Home And Polyclinic', 'Jain Sant Phool Chand Ji Charitable Hospital', 'Sarvodya Hospital', 'Bhatnagar Maternity and Nursing Home', 'Tirath Ram Hospitals Pvt L

In [246]:
# Converting price into 'cr' and 'float' datatype

def treat_price(x):
    if type(x) == float:
        return x
    else:
        if x[1] == 'Lac':
            return round(float(x[0])/100,2)
        else:
            return round(float(x[0]),2)

In [247]:
df['price'] = df['price'].str.split(' ').apply(treat_price)

In [248]:
df.head()

property_name                 society  price  \
0     2 BHK Flat in Krishna Colony  maa bhagwati residency   0.45   
1        2 BHK Flat in Ashok Vihar            apna enclave   0.50   
2              2 BHK Flat in Sohna  tulsiani easy in homes   0.40   
3  2 BHK Flat in Sector 61 Gurgaon     smart world orchard   1.47   
4  2 BHK Flat in Sector 92 Gurgaon        parkwood westend   0.70   

    price_per_sqft                            areaWithType     bedRoom  \
0   ₹ 5,000/sq.ft.          Carpet area: 900 (83.61 sq.m.)  2 Bedrooms   
1   ₹ 7,692/sq.ft.          Carpet area: 650 (60.39 sq.m.)  2 Bedrooms   
2   ₹ 6,722/sq.ft.          Carpet area: 595 (55.28 sq.m.)  2 Bedrooms   
3  ₹ 12,250/sq.ft.        Carpet area: 1200 (111.48 sq.m.)  2 Bedrooms   
4   ₹ 5,204/sq.ft.  Super Built up area 1345(124.95 sq.m.)  2 Bedrooms   

      bathroom      balcony additionalRoom  \
0  2 Bathrooms    1 Balcony            NaN   
1  2 Bathrooms    1 Balcony            NaN   
2  2 Bathrooms  3 Balconies            NaN   
3  2 Bathrooms  2 Balconies     Study Room   
4  2 Bathrooms  3 Balconies     Study Room   

                               address             floorNum facing  \
0     Krishna Colony, Gurgaon, Haryana    4th   of 4 Floors   West   
1   46b, Ashok Vihar, Gurgaon, Haryana    1st   of 3 Floors   West   
2              Sohna, Gurgaon, Haryana  12nd   of 14 Floors    NaN   
3  Sector 61 Gurgaon, Gurgaon, Haryana    2nd   of 4 Floors    NaN   
4  Sector 92 Gurgaon, Gurgaon, Haryana    5th   of 8 Floors    NaN   

        agePossession  \
0     1 to 5 Year Old   
1        10+ Year Old   
2     0 to 1 Year Old   
3            Dec 2023   
4  Under Construction   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                nearbyLocations  \
0  ['Chintapurni Mandir', 'State bank ATM', 'Pearl Dental Clinic', 'Taneja Hospital', 'Dr. Ashok Jain', 'Dr. Madan Clinic', 'Geeta Nursing Home Gurgaon', 'Chiranjiv Hospital', 'Dev Man Kathuria Clinic', 'Aryan Hospital', 'Kr Dental Hub', 'Swastik Maternity and Medical Centre', 'Satyam Hospital Gurgaon', 'Ravi Clinic and Health Care Centre', 'D.R. Rajnis Gupta Clinic', 'Esic Hospital Gurugram', 'Gurgaon Eye Centre', 'Bindal Clinic', 'Shiv Mahima Patient Care Bureau', 'Shree Krishna Hospital Gurgaon', 'M.S Hospital', 'My Care Clinic', 'Ankur Clinic and Maternity Home', 'Triveni Hospital Gurgaon', 'Shri Gobind Hospital', 'Navjeevan Hospital and Maternity Centre', 'Lal Superspeciality Hospital', 'Sparsh Hospital Gurgaon', 'Prateek Nursing Home And Polyclinic', 'Jain Sant Phool Chand Ji Charitable Hospital', 'Sarvodya Hospital', 'Bhatnagar Maternity and Nursing Home', 'Tirath Ram Hospitals Pvt Ltd', 'Dr. Sarvejeet Singh', 'C

# Price Per Sqft

In [249]:
df['price_per_sqft']

0         ₹ 5,000/sq.ft.
1         ₹ 7,692/sq.ft.
2         ₹ 6,722/sq.ft.
3        ₹ 12,250/sq.ft.
4         ₹ 5,204/sq.ft.
5         ₹ 6,269/sq.ft.
6        ₹ 13,333/sq.ft.
7         ₹ 7,860/sq.ft.
8         ₹ 8,148/sq.ft.
9        ₹ 16,885/sq.ft.
10        ₹ 9,767/sq.ft.
11        ₹ 5,587/sq.ft.
12        ₹ 6,940/sq.ft.
13        ₹ 6,859/sq.ft.
14        ₹ 5,384/sq.ft.
15       ₹ 16,859/sq.ft.
16       ₹ 14,084/sq.ft.
17       ₹ 11,475/sq.ft.
18                   NaN
19       ₹ 10,476/sq.ft.
20        ₹ 9,156/sq.ft.
21        ₹ 8,604/sq.ft.
22        ₹ 7,796/sq.ft.
23        ₹ 9,135/sq.ft.
24       ₹ 12,855/sq.ft.
26        ₹ 7,910/sq.ft.
27       ₹ 18,331/sq.ft.
28       ₹ 24,300/sq.ft.
29        ₹ 5,600/sq.ft.
30       ₹ 12,784/sq.ft.
31       ₹ 26,936/sq.ft.
32       ₹ 10,156/sq.ft.
33        ₹ 7,194/sq.ft.
34        ₹ 6,538/sq.ft.
35        ₹ 5,769/sq.ft.
36        ₹ 6,950/sq.ft.
37        ₹ 8,866/sq.ft.
38        ₹ 5,403/sq.ft.
39        ₹ 5,714/sq.ft.
40       ₹ 15,151/sq.ft.


### Observations:- 

- In each flat listing, the price is expressed per square foot, a crucial metic in Real State.
- This data is continuous, and i am interested in extracting a single numerical value from this column.
- The format is like ₹ 12,500/sq.ft. 
- To process this, I plan to remove the rupee symbol, commas from numbers, and the "/sq.ft" suffix.
- Initially, I will examine the different categories present in this column using the value_counts function.

In [250]:
df['price_per_sqft'].value_counts()

price_per_sqft
₹ 10,000/sq.ft.     19
₹ 8,000/sq.ft.      16
₹ 12,500/sq.ft.     16
₹ 6,666/sq.ft.      13
₹ 5,000/sq.ft.      13
₹ 7,500/sq.ft.      12
₹ 8,333/sq.ft.      12
₹ 6,000/sq.ft.      11
₹ 8,461/sq.ft.       9
₹ 12,000/sq.ft.      8
₹ 7,000/sq.ft.       8
₹ 9,000/sq.ft.       7
₹ 9,230/sq.ft.       6
₹ 8,928/sq.ft.       6
₹ 5,500/sq.ft.       6
₹ 11,500/sq.ft.      6
₹ 11,111/sq.ft.      6
₹ 6,578/sq.ft.       6
₹ 8,888/sq.ft.       6
₹ 8,205/sq.ft.       5
₹ 7,692/sq.ft.       5
₹ 8,571/sq.ft.       5
₹ 4,666/sq.ft.       5
₹ 5,556/sq.ft.       5
₹ 4,615/sq.ft.       5
₹ 7,641/sq.ft.       5
₹ 5,384/sq.ft.       5
₹ 14,242/sq.ft.      5
₹ 7,142/sq.ft.       5
₹ 10,714/sq.ft.      5
₹ 4,444/sq.ft.       5
₹ 5,600/sq.ft.       5
₹ 7,407/sq.ft.       5
₹ 6,500/sq.ft.       5
₹ 11,428/sq.ft.      5
₹ 16,000/sq.ft.      5
₹ 7,250/sq.ft.       4
₹ 9,822/sq.ft.       4
₹ 13,913/sq.ft.      4
₹ 5,660/sq.ft.       4
₹ 8,043/sq.ft.       4
₹ 8,500/sq.ft.       4
₹ 9,210/sq.ft.     

- I observed that the entire dataset maintains consistency, with prices consistently provided in terms of square feet.
- Consequently, I will proceed with further processing based on the standardized format.

In [140]:
df['price_per_sqft'].str.split('/').str.get(0).str.replace('₹', '').str.replace(',','').str.strip().astype('float')

0         5000.0
1         7692.0
2         6722.0
3        12250.0
4         5204.0
5         6269.0
6        13333.0
7         7860.0
8         8148.0
9        16885.0
10        9767.0
11        5587.0
12        6940.0
13        6859.0
14        5384.0
15       16859.0
16       14084.0
17       11475.0
18           NaN
19       10476.0
20        9156.0
21        8604.0
22        7796.0
23        9135.0
24       12855.0
26        7910.0
27       18331.0
28       24300.0
29        5600.0
30       12784.0
31       26936.0
32       10156.0
33        7194.0
34        6538.0
35        5769.0
36        6950.0
37        8866.0
38        5403.0
39        5714.0
40       15151.0
41        6200.0
42       16000.0
43        6378.0
44        3571.0
45        7846.0
46        7297.0
47        7581.0
48       10005.0
49       10714.0
50       16969.0
51       14545.0
52        5424.0
53        7410.0
54        6400.0
55        4715.0
56        5000.0
57        6666.0
58        9872.0
59        6594

In [251]:
df['price_per_sqft'] = df['price_per_sqft'].str.split('/').str.get(0).str.replace('₹', '').str.replace(',','').str.strip().astype('float')

In [252]:
df.head()

property_name                 society  price  \
0     2 BHK Flat in Krishna Colony  maa bhagwati residency   0.45   
1        2 BHK Flat in Ashok Vihar            apna enclave   0.50   
2              2 BHK Flat in Sohna  tulsiani easy in homes   0.40   
3  2 BHK Flat in Sector 61 Gurgaon     smart world orchard   1.47   
4  2 BHK Flat in Sector 92 Gurgaon        parkwood westend   0.70   

   price_per_sqft                            areaWithType     bedRoom  \
0          5000.0          Carpet area: 900 (83.61 sq.m.)  2 Bedrooms   
1          7692.0          Carpet area: 650 (60.39 sq.m.)  2 Bedrooms   
2          6722.0          Carpet area: 595 (55.28 sq.m.)  2 Bedrooms   
3         12250.0        Carpet area: 1200 (111.48 sq.m.)  2 Bedrooms   
4          5204.0  Super Built up area 1345(124.95 sq.m.)  2 Bedrooms   

      bathroom      balcony additionalRoom  \
0  2 Bathrooms    1 Balcony            NaN   
1  2 Bathrooms    1 Balcony            NaN   
2  2 Bathrooms  3 Balconies            NaN   
3  2 Bathrooms  2 Balconies     Study Room   
4  2 Bathrooms  3 Balconies     Study Room   

                               address             floorNum facing  \
0     Krishna Colony, Gurgaon, Haryana    4th   of 4 Floors   West   
1   46b, Ashok Vihar, Gurgaon, Haryana    1st   of 3 Floors   West   
2              Sohna, Gurgaon, Haryana  12nd   of 14 Floors    NaN   
3  Sector 61 Gurgaon, Gurgaon, Haryana    2nd   of 4 Floors    NaN   
4  Sector 92 Gurgaon, Gurgaon, Haryana    5th   of 8 Floors    NaN   

        agePossession  \
0     1 to 5 Year Old   
1        10+ Year Old   
2     0 to 1 Year Old   
3            Dec 2023   
4  Under Construction   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                nearbyLocations  \
0  ['Chintapurni Mandir', 'State bank ATM', 'Pearl Dental Clinic', 'Taneja Hospital', 'Dr. Ashok Jain', 'Dr. Madan Clinic', 'Geeta Nursing Home Gurgaon', 'Chiranjiv Hospital', 'Dev Man Kathuria Clinic', 'Aryan Hospital', 'Kr Dental Hub', 'Swastik Maternity and Medical Centre', 'Satyam Hospital Gurgaon', 'Ravi Clinic and Health Care Centre', 'D.R. Rajnis Gupta Clinic', 'Esic Hospital Gurugram', 'Gurgaon Eye Centre', 'Bindal Clinic', 'Shiv Mahima Patient Care Bureau', 'Shree Krishna Hospital Gurgaon', 'M.S Hospital', 'My Care Clinic', 'Ankur Clinic and Maternity Home', 'Triveni Hospital Gurgaon', 'Shri Gobind Hospital', 'Navjeevan Hospital and Maternity Centre', 'Lal Superspeciality Hospital', 'Sparsh Hospital Gurgaon', 'Prateek Nursing Home And Polyclinic', 'Jain Sant Phool Chand Ji Charitable Hospital', 'Sarvodya Hospital', 'Bhatnagar Maternity and Nursing Home', 'Tirath Ram Hospitals Pvt Ltd', 'Dr. Sarvejeet Singh', 'Centre 

# Bedroom

In [253]:
df['bedRoom']

0       2 Bedrooms
1       2 Bedrooms
2       2 Bedrooms
3       2 Bedrooms
4       2 Bedrooms
5       2 Bedrooms
6       3 Bedrooms
7       3 Bedrooms
8       2 Bedrooms
9       3 Bedrooms
10      2 Bedrooms
11      2 Bedrooms
12      3 Bedrooms
13      2 Bedrooms
14      2 Bedrooms
15      4 Bedrooms
16       1 Bedroom
17      2 Bedrooms
18      2 Bedrooms
19      3 Bedrooms
20      2 Bedrooms
21      3 Bedrooms
22      2 Bedrooms
23      3 Bedrooms
24      3 Bedrooms
26      2 Bedrooms
27      3 Bedrooms
28      3 Bedrooms
29      2 Bedrooms
30      2 Bedrooms
31      2 Bedrooms
32      4 Bedrooms
33      2 Bedrooms
34      3 Bedrooms
35      2 Bedrooms
36      2 Bedrooms
37      3 Bedrooms
38      2 Bedrooms
39      2 Bedrooms
40      3 Bedrooms
41      2 Bedrooms
42      3 Bedrooms
43      4 Bedrooms
44      2 Bedrooms
45      3 Bedrooms
46      4 Bedrooms
47      2 Bedrooms
48      3 Bedrooms
49      3 Bedrooms
50      3 Bedrooms
51      2 Bedrooms
52      2 Bedrooms
53      3 Be

In [254]:
df['bedRoom'].value_counts()

bedRoom
3 Bedrooms    1437
2 Bedrooms     944
4 Bedrooms     478
1 Bedroom      104
5 Bedrooms      31
6 Bedrooms       3
Name: count, dtype: int64

In [255]:
# checking Null Values

df[df['bedRoom'].isnull()]

property_name                       society  \
2849           2 BHK Flat in Sector 107 Gurgaon       signature global solera   
2850            4 BHK Flat in Sector 53 Gurgaon                tulip monsella   
2851              2 BHK Flat in New Palam Vihar                       my home   
2852                        2 BHK Flat in Sohna        breez global hill view   
2922           3 BHK Flat in Sector 99A Gurgaon      pareena coban residences   
2923         1 BHK Flat in Golf Course Ext Road  ikon tower baani city centre   
2924            2 BHK Flat in Sector 89 Gurgaon                   greenopolis   
2926  4 BHK Flat in C Block Sushant Lok Phase 1                 maple heights   
2927            2 BHK Flat in Sector 99 Gurgaon                assotech blith   

      price  price_per_sqft areaWithType bedRoom bathroom balcony  \
2849    NaN             NaN          NaN     NaN      NaN     NaN   
2850    NaN         33198.0          NaN     NaN      NaN     NaN   
2851    NaN          4400.0          NaN     NaN      NaN     NaN   
2852    NaN          5470.0          NaN     NaN      NaN     NaN   
2922    NaN          5759.0          NaN     NaN      NaN     NaN   
2923    NaN         12437.0          NaN     NaN      NaN     NaN   
2924    NaN          4820.0          NaN     NaN      NaN     NaN   
2926    NaN          8888.0          NaN     NaN      NaN     NaN   
2927    NaN          6593.0          NaN     NaN      NaN     NaN   

     additionalRoom address floorNum facing agePossession nearbyLocations  \
2849            NaN     NaN      NaN    NaN           NaN             NaN   
2850            NaN     NaN      NaN    NaN           NaN             NaN   
2851            NaN     NaN      NaN    NaN           NaN             NaN   
2852            NaN     NaN      NaN    NaN           NaN             NaN   
2922            NaN     NaN      NaN    NaN           NaN             NaN   
2923            NaN     NaN      NaN    NaN           NaN             NaN   
2924            NaN     NaN      NaN    NaN           NaN             NaN   
2926            NaN     NaN      NaN    NaN           NaN             NaN   
2927            NaN     NaN      NaN    NaN           NaN             NaN   

     description furnishDetails features rating  
2849         NaN            NaN      NaN    NaN  
2850         NaN            NaN      NaN    NaN  
2851         NaN            NaN      NaN    NaN  
2852         NaN            NaN      NaN    NaN  
2922         NaN            NaN      NaN    NaN  
2923         NaN            NaN      NaN    NaN  
2924         NaN            NaN      NaN    NaN  
2926         NaN            NaN      NaN    NaN  
2927         NaN            NaN      NaN    NaN

### Observations:-
- I identified that in 9 flats, null value were present not just in the bedroom column but acroos all columns.
- Predicting value with this amount of missing data may not be reliable.
- Therfore, I have decided to remove these rows,selecting values where the bedroom is not null for further analysis.

In [256]:
df = df[~df['bedRoom'].isnull()]

In [257]:
df['bedRoom'].isnull().sum()

0

- all Null values now removed.

In [258]:
df['bedRoom']= df['bedRoom'].str.split(' ').str.get(0).astype('int')

In [259]:
df.head()

property_name                 society  price  \
0     2 BHK Flat in Krishna Colony  maa bhagwati residency   0.45   
1        2 BHK Flat in Ashok Vihar            apna enclave   0.50   
2              2 BHK Flat in Sohna  tulsiani easy in homes   0.40   
3  2 BHK Flat in Sector 61 Gurgaon     smart world orchard   1.47   
4  2 BHK Flat in Sector 92 Gurgaon        parkwood westend   0.70   

   price_per_sqft                            areaWithType  bedRoom  \
0          5000.0          Carpet area: 900 (83.61 sq.m.)        2   
1          7692.0          Carpet area: 650 (60.39 sq.m.)        2   
2          6722.0          Carpet area: 595 (55.28 sq.m.)        2   
3         12250.0        Carpet area: 1200 (111.48 sq.m.)        2   
4          5204.0  Super Built up area 1345(124.95 sq.m.)        2   

      bathroom      balcony additionalRoom  \
0  2 Bathrooms    1 Balcony            NaN   
1  2 Bathrooms    1 Balcony            NaN   
2  2 Bathrooms  3 Balconies            NaN   
3  2 Bathrooms  2 Balconies     Study Room   
4  2 Bathrooms  3 Balconies     Study Room   

                               address             floorNum facing  \
0     Krishna Colony, Gurgaon, Haryana    4th   of 4 Floors   West   
1   46b, Ashok Vihar, Gurgaon, Haryana    1st   of 3 Floors   West   
2              Sohna, Gurgaon, Haryana  12nd   of 14 Floors    NaN   
3  Sector 61 Gurgaon, Gurgaon, Haryana    2nd   of 4 Floors    NaN   
4  Sector 92 Gurgaon, Gurgaon, Haryana    5th   of 8 Floors    NaN   

        agePossession  \
0     1 to 5 Year Old   
1        10+ Year Old   
2     0 to 1 Year Old   
3            Dec 2023   
4  Under Construction   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                nearbyLocations  \
0  ['Chintapurni Mandir', 'State bank ATM', 'Pearl Dental Clinic', 'Taneja Hospital', 'Dr. Ashok Jain', 'Dr. Madan Clinic', 'Geeta Nursing Home Gurgaon', 'Chiranjiv Hospital', 'Dev Man Kathuria Clinic', 'Aryan Hospital', 'Kr Dental Hub', 'Swastik Maternity and Medical Centre', 'Satyam Hospital Gurgaon', 'Ravi Clinic and Health Care Centre', 'D.R. Rajnis Gupta Clinic', 'Esic Hospital Gurugram', 'Gurgaon Eye Centre', 'Bindal Clinic', 'Shiv Mahima Patient Care Bureau', 'Shree Krishna Hospital Gurgaon', 'M.S Hospital', 'My Care Clinic', 'Ankur Clinic and Maternity Home', 'Triveni Hospital Gurgaon', 'Shri Gobind Hospital', 'Navjeevan Hospital and Maternity Centre', 'Lal Superspeciality Hospital', 'Sparsh Hospital Gurgaon', 'Prateek Nursing Home And Polyclinic', 'Jain Sant Phool Chand Ji Charitable Hospital', 'Sarvodya Hospital', 'Bhatnagar Maternity and Nursing Home', 'Tirath Ram Hospitals Pvt Ltd', 'Dr. Sarvejeet Singh', 'Centre For Sight Gurgaon 

# Bathrooms

In [260]:
df['bathroom']

0       2 Bathrooms
1       2 Bathrooms
2       2 Bathrooms
3       2 Bathrooms
4       2 Bathrooms
5       2 Bathrooms
6       3 Bathrooms
7       4 Bathrooms
8       4 Bathrooms
9       3 Bathrooms
10      2 Bathrooms
11      2 Bathrooms
12      3 Bathrooms
13      2 Bathrooms
14      2 Bathrooms
15      5 Bathrooms
16       1 Bathroom
17      2 Bathrooms
18      2 Bathrooms
19      3 Bathrooms
20      2 Bathrooms
21      4 Bathrooms
22      3 Bathrooms
23      3 Bathrooms
24      3 Bathrooms
26      2 Bathrooms
27      4 Bathrooms
28      5 Bathrooms
29      2 Bathrooms
30      2 Bathrooms
31      2 Bathrooms
32      4 Bathrooms
33      2 Bathrooms
34      2 Bathrooms
35      2 Bathrooms
36      2 Bathrooms
37      3 Bathrooms
38      2 Bathrooms
39      2 Bathrooms
40      3 Bathrooms
41      2 Bathrooms
42      3 Bathrooms
43      4 Bathrooms
44      2 Bathrooms
45      2 Bathrooms
46      4 Bathrooms
47      2 Bathrooms
48      3 Bathrooms
49      3 Bathrooms
50      3 Bathrooms


In [261]:
df['bathroom'].isnull().sum() # no missing value

0

In [262]:
df['bathroom'].value_counts()

bathroom
2 Bathrooms    1044
3 Bathrooms     989
4 Bathrooms     636
5 Bathrooms     169
1 Bathroom      112
6 Bathrooms      42
7 Bathrooms       5
Name: count, dtype: int64

In [263]:
df['bathroom'].str.split(' ').str.get(0).astype('string')

0       2
1       2
2       2
3       2
4       2
5       2
6       3
7       4
8       4
9       3
10      2
11      2
12      3
13      2
14      2
15      5
16      1
17      2
18      2
19      3
20      2
21      4
22      3
23      3
24      3
26      2
27      4
28      5
29      2
30      2
31      2
32      4
33      2
34      2
35      2
36      2
37      3
38      2
39      2
40      3
41      2
42      3
43      4
44      2
45      2
46      4
47      2
48      3
49      3
50      3
51      2
52      2
53      2
54      3
55      4
56      1
57      1
58      4
59      2
60      2
61      2
62      5
63      2
64      3
65      4
66      3
67      3
68      2
69      2
70      2
71      2
72      2
73      2
74      3
75      3
76      2
77      2
78      4
79      4
80      2
81      3
82      2
83      5
84      2
85      2
86      4
87      3
88      2
89      1
90      2
91      2
92      2
93      2
94      3
95      2
96      4
97      3
98      2
99      3
100     3


In [264]:
df['bathroom'] = df['bathroom'].str.split(' ').str.get(0).astype('string')

In [265]:
df.head()

property_name                 society  price  \
0     2 BHK Flat in Krishna Colony  maa bhagwati residency   0.45   
1        2 BHK Flat in Ashok Vihar            apna enclave   0.50   
2              2 BHK Flat in Sohna  tulsiani easy in homes   0.40   
3  2 BHK Flat in Sector 61 Gurgaon     smart world orchard   1.47   
4  2 BHK Flat in Sector 92 Gurgaon        parkwood westend   0.70   

   price_per_sqft                            areaWithType  bedRoom bathroom  \
0          5000.0          Carpet area: 900 (83.61 sq.m.)        2        2   
1          7692.0          Carpet area: 650 (60.39 sq.m.)        2        2   
2          6722.0          Carpet area: 595 (55.28 sq.m.)        2        2   
3         12250.0        Carpet area: 1200 (111.48 sq.m.)        2        2   
4          5204.0  Super Built up area 1345(124.95 sq.m.)        2        2   

       balcony additionalRoom                              address  \
0    1 Balcony            NaN     Krishna Colony, Gurgaon, Haryana   
1    1 Balcony            NaN   46b, Ashok Vihar, Gurgaon, Haryana   
2  3 Balconies            NaN              Sohna, Gurgaon, Haryana   
3  2 Balconies     Study Room  Sector 61 Gurgaon, Gurgaon, Haryana   
4  3 Balconies     Study Room  Sector 92 Gurgaon, Gurgaon, Haryana   

              floorNum facing       agePossession  \
0    4th   of 4 Floors   West     1 to 5 Year Old   
1    1st   of 3 Floors   West        10+ Year Old   
2  12nd   of 14 Floors    NaN     0 to 1 Year Old   
3    2nd   of 4 Floors    NaN            Dec 2023   
4    5th   of 8 Floors    NaN  Under Construction   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                nearbyLocations  \
0  ['Chintapurni Mandir', 'State bank ATM', 'Pearl Dental Clinic', 'Taneja Hospital', 'Dr. Ashok Jain', 'Dr. Madan Clinic', 'Geeta Nursing Home Gurgaon', 'Chiranjiv Hospital', 'Dev Man Kathuria Clinic', 'Aryan Hospital', 'Kr Dental Hub', 'Swastik Maternity and Medical Centre', 'Satyam Hospital Gurgaon', 'Ravi Clinic and Health Care Centre', 'D.R. Rajnis Gupta Clinic', 'Esic Hospital Gurugram', 'Gurgaon Eye Centre', 'Bindal Clinic', 'Shiv Mahima Patient Care Bureau', 'Shree Krishna Hospital Gurgaon', 'M.S Hospital', 'My Care Clinic', 'Ankur Clinic and Maternity Home', 'Triveni Hospital Gurgaon', 'Shri Gobind Hospital', 'Navjeevan Hospital and Maternity Centre', 'Lal Superspeciality Hospital', 'Sparsh Hospital Gurgaon', 'Prateek Nursing Home And Polyclinic', 'Jain Sant Phool Chand Ji Charitable Hospital', 'Sarvodya Hospital', 'Bhatnagar Maternity and Nursing Home', 'Tirath Ram Hospitals Pvt Ltd', 'Dr. Sarvejeet Singh', 'Centre For Sight Gurgaon New Railway Road', 'Dr. Ajay S. Gupta Clinic', 'Kidney 

# Balcony

In [266]:
df['balcony']

0          1 Balcony
1          1 Balcony
2        3 Balconies
3        2 Balconies
4        3 Balconies
5        3 Balconies
6        3 Balconies
7        3 Balconies
8       3+ Balconies
9        2 Balconies
10       2 Balconies
11         1 Balcony
12       3 Balconies
13      3+ Balconies
14       3 Balconies
15       3 Balconies
16        No Balcony
17       2 Balconies
18       3 Balconies
19       3 Balconies
20       2 Balconies
21      3+ Balconies
22       3 Balconies
23      3+ Balconies
24      3+ Balconies
26         1 Balcony
27       2 Balconies
28      3+ Balconies
29       2 Balconies
30        No Balcony
31       2 Balconies
32      3+ Balconies
33         1 Balcony
34       2 Balconies
35       3 Balconies
36      3+ Balconies
37       2 Balconies
38       2 Balconies
39       3 Balconies
40       2 Balconies
41      3+ Balconies
42      3+ Balconies
43      3+ Balconies
44       2 Balconies
45      3+ Balconies
46      3+ Balconies
47       2 Balconies
48      3+ Ba

### Observations:-
- some rows have value like "No Balcony" and some place value given like " 3+". so it could be any number.

In [267]:
df['balcony'].value_counts()

balcony
3 Balconies     974
3+ Balconies    862
2 Balconies     749
1 Balcony       315
No Balcony       97
Name: count, dtype: int64

In [268]:
df['balcony'].isnull().sum() # no missing values

0

- I replaced instances of "No Balcony" with "0".
- I chose not to apply the astype function to convert the data into integers. The reason is that we have categories 
like "3+balcony", where the number of balconies can be 4,5,6. Therfore, it is more appropriate to treat the balcony
count as a string rather than a single numeric value.

In [269]:
df['balcony'].str.split(' ').str.get(0).str.replace('No','0')

0        1
1        1
2        3
3        2
4        3
5        3
6        3
7        3
8       3+
9        2
10       2
11       1
12       3
13      3+
14       3
15       3
16       0
17       2
18       3
19       3
20       2
21      3+
22       3
23      3+
24      3+
26       1
27       2
28      3+
29       2
30       0
31       2
32      3+
33       1
34       2
35       3
36      3+
37       2
38       2
39       3
40       2
41      3+
42      3+
43      3+
44       2
45      3+
46      3+
47       2
48      3+
49      3+
50      3+
51       2
52       3
53       2
54       2
55      3+
56       1
57       1
58       1
59       2
60       2
61       3
62      3+
63       2
64      3+
65       3
66      3+
67       3
68       2
69       2
70       3
71       3
72       2
73       1
74       3
75       2
76       2
77       3
78      3+
79       3
80       2
81       0
82       1
83      3+
84       1
85       1
86      3+
87       1
88       1
89       1
90       2
91       2

In [270]:
df['balcony'] = df['balcony'].str.split(' ').str.get(0).str.replace('No','0')

In [271]:
df.head()

property_name                 society  price  \
0     2 BHK Flat in Krishna Colony  maa bhagwati residency   0.45   
1        2 BHK Flat in Ashok Vihar            apna enclave   0.50   
2              2 BHK Flat in Sohna  tulsiani easy in homes   0.40   
3  2 BHK Flat in Sector 61 Gurgaon     smart world orchard   1.47   
4  2 BHK Flat in Sector 92 Gurgaon        parkwood westend   0.70   

   price_per_sqft                            areaWithType  bedRoom bathroom  \
0          5000.0          Carpet area: 900 (83.61 sq.m.)        2        2   
1          7692.0          Carpet area: 650 (60.39 sq.m.)        2        2   
2          6722.0          Carpet area: 595 (55.28 sq.m.)        2        2   
3         12250.0        Carpet area: 1200 (111.48 sq.m.)        2        2   
4          5204.0  Super Built up area 1345(124.95 sq.m.)        2        2   

  balcony additionalRoom                              address  \
0       1            NaN     Krishna Colony, Gurgaon, Haryana   
1       1            NaN   46b, Ashok Vihar, Gurgaon, Haryana   
2       3            NaN              Sohna, Gurgaon, Haryana   
3       2     Study Room  Sector 61 Gurgaon, Gurgaon, Haryana   
4       3     Study Room  Sector 92 Gurgaon, Gurgaon, Haryana   

              floorNum facing       agePossession  \
0    4th   of 4 Floors   West     1 to 5 Year Old   
1    1st   of 3 Floors   West        10+ Year Old   
2  12nd   of 14 Floors    NaN     0 to 1 Year Old   
3    2nd   of 4 Floors    NaN            Dec 2023   
4    5th   of 8 Floors    NaN  Under Construction   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                nearbyLocations  \
0  ['Chintapurni Mandir', 'State bank ATM', 'Pearl Dental Clinic', 'Taneja Hospital', 'Dr. Ashok Jain', 'Dr. Madan Clinic', 'Geeta Nursing Home Gurgaon', 'Chiranjiv Hospital', 'Dev Man Kathuria Clinic', 'Aryan Hospital', 'Kr Dental Hub', 'Swastik Maternity and Medical Centre', 'Satyam Hospital Gurgaon', 'Ravi Clinic and Health Care Centre', 'D.R. Rajnis Gupta Clinic', 'Esic Hospital Gurugram', 'Gurgaon Eye Centre', 'Bindal Clinic', 'Shiv Mahima Patient Care Bureau', 'Shree Krishna Hospital Gurgaon', 'M.S Hospital', 'My Care Clinic', 'Ankur Clinic and Maternity Home', 'Triveni Hospital Gurgaon', 'Shri Gobind Hospital', 'Navjeevan Hospital and Maternity Centre', 'Lal Superspeciality Hospital', 'Sparsh Hospital Gurgaon', 'Prateek Nursing Home And Polyclinic', 'Jain Sant Phool Chand Ji Charitable Hospital', 'Sarvodya Hospital', 'Bhatnagar Maternity and Nursing Home', 'Tirath Ram Hospitals Pvt Ltd', 'Dr. Sarvejeet Singh', 'Centre For Sight Gurgaon New Railway Road', 'Dr. Ajay S. Gupta Clinic', 'Kidney Clinic', 'Mangalam Hospital an

# Additional Room

In [272]:
df['additionalRoom'].value_counts()

additionalRoom
Servant Room                                     629
Study Room                                       232
Others                                           179
Pooja Room                                       132
Study Room,Servant Room                           81
Store Room                                        76
Pooja Room,Servant Room                           60
Servant Room,Others                               52
Servant Room,Pooja Room                           30
Study Room,Others                                 27
Pooja Room,Study Room,Servant Room,Others         25
Pooja Room,Study Room,Servant Room                24
Servant Room,Store Room                           19
Pooja Room,Study Room                             13
Pooja Room,Study Room,Servant Room,Store Room     12
Study Room,Pooja Room                              8
Servant Room,Study Room                            8
Study Room,Servant Room,Store Room                 7
Pooja Room,Store Room          

In [273]:
df['additionalRoom'].isnull().sum()

1305

In [274]:
df['additionalRoom'].value_counts().shape

(49,)

### Observations:-

- There are total of 49 categories specified in the "additional room"column.
- Among the entire dataset, 1305 flats have Zero additional rooms which represents as "Nan"
- To enhance clarity,I will replace Null values "Nan" in this column with "not available", distingushing between flats
with additional rooms and those without.
- Additionally, I will convert all values to lowercase for consistency.

In [275]:
# filling Nan value with "not available"
df['additionalRoom'].fillna('not available', inplace = True)

In [276]:
df['additionalRoom']= df['additionalRoom'].str.lower()

In [277]:
df['additionalRoom']

0                                       not available
1                                       not available
2                                       not available
3                                          study room
4                                          study room
5                                       not available
6                                       not available
7                                       not available
8                                          study room
9                                        servant room
10                                      not available
11                                      not available
12                                      not available
13                                      not available
14                                         study room
15                                      not available
16                                      not available
17                                      not available
18                          

In [278]:
df.head(1)

property_name                 society  price  \
0  2 BHK Flat in Krishna Colony  maa bhagwati residency   0.45   

   price_per_sqft                    areaWithType  bedRoom bathroom balcony  \
0          5000.0  Carpet area: 900 (83.61 sq.m.)        2        2       1   

  additionalRoom                           address           floorNum facing  \
0  not available  Krishna Colony, Gurgaon, Haryana  4th   of 4 Floors   West   

     agePossession  \
0  1 to 5 Year Old   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                nearbyLocations  \
0  ['Chintapurni Mandir', 'State bank ATM', 'Pearl Dental Clinic', 'Taneja Hospital', 'Dr. Ashok Jain', 'Dr. Madan Clinic', 'Geeta Nursing Home Gurgaon', 'Chiranjiv Hospital', 'Dev Man Kathuria Clinic', 'Aryan Hospital', 'Kr Dental Hub', 'Swastik Maternity and Medical Centre', 'Satyam Hospital Gurgaon', 'Ravi Clinic and Health Care Centre', 'D.R. Rajnis Gupta Clinic', 'Esic Hospital Gurugram', 'Gurgaon Eye Centre', 'Bindal Clinic', 'Shiv Mahima Patient Care Bureau', 'Shree Krishna Hospital Gurgaon', 'M.S Hospital', 'My Care Clinic', 'Ankur Clinic and Maternity Home', 'Triveni Hospital Gurgaon', 'Shri Gobind Hospital', 'Navjeevan Hospital and Maternity Centre', 'Lal Superspeciality Hospital', 'Sparsh Hospital Gurgaon', 'Prateek Nursing Home And Polyclinic', 'Jain Sant Phool Chand Ji Charitable Hospital', 'Sarvodya Hospital', 'Bhatnagar Maternity and Nursing Home', 'Tirath Ram Hospitals Pvt Ltd', 'Dr. Sarvejeet Singh', 'Centre For Sight Gurgaon New Railway Road', 'Dr. Ajay S. Gupta Clinic', 'Kidney Clinic', 'Mangalam Hospital and Heart Centre Gurgaon', 'Sethi Hospital Gurgaon', 'Shubham Hospital Gurgaon', 'Kharbanda Maternity and Nursing Home', 'Gupta Hospital Gurgaon', 'Dentecare - Multispeciality Dental Clinic', 'State bank of india', 'Hdfc bank', 'Kotak bank', 'Indian bank', 'Pizza Hut', 'St. Michaels Sr. Sec. School', 'Dronacharya Government College']   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  description  \
0  So with lift.Maa bhagwati residency is one of gurgaon's most sought after destination for apartments and this 2 bhk flat in krishna colony is your opportunity to be a part of this community. The flat occupies a carpet area of 900 sq.Ft. That consists of 2 bedrooms, 2 bathrooms

# Floor Number

In [279]:
df['floorNum'].value_counts()

floorNum
2nd   of 4 Floors           74
3rd   of 4 Floors           71
4th   of 4 Floors           62
1st   of 4 Floors           61
12nd   of 14 Floors         49
14th   of 14 Floors         48
Ground of 14 Floors         40
10th   of 14 Floors         35
7th   of 14 Floors          35
8th   of 14 Floors          34
4th   of 14 Floors          28
6th   of 14 Floors          27
2nd   of 2 Floors           26
1st   of 14 Floors          26
3rd   of 3 Floors           26
3rd   of 14 Floors          24
5th   of 14 Floors          24
8th   of 19 Floors          24
11st   of 14 Floors         23
1st   of 1 Floors           23
9th   of 14 Floors          23
9th   of 9 Floors           23
5th   of 12 Floors          22
2nd   of 3 Floors           22
2nd   of 14 Floors          21
8th   of 18 Floors          20
10th   of 19 Floors         18
6th   of 18 Floors          18
10th   of 18 Floors         17
9th   of 13 Floors          17
7th   of 15 Floors          17
12nd   of 12 Floors         17

In [280]:
df['floorNum'].isnull().sum()

2

In [281]:
df[df['floorNum'].isnull()] 

# for index number '181' floor number value can fill by taking value from address.

property_name             society  price  \
181   3 BHK Flat in Dwarka Expressway Gurgaon  experion heartsong   1.08   
2766          2 BHK Flat in Sector 78 Gurgaon                 nan   0.60   

      price_per_sqft                        areaWithType  bedRoom bathroom  \
181           6150.0  Built Up area: 1758 (163.32 sq.m.)        3        3   
2766          3692.0  Built Up area: 1625 (150.97 sq.m.)        2        2   

     balcony additionalRoom  \
181        0  not available   
2766       0  not available   

                                                                                address  \
181   604, Tower B-3, 6th Floor,Sector 108, Dwarka Expressway Gurgaon, Gurgaon, Haryana   
2766                                       Gurgaon, Sector 78 Gurgaon, Gurgaon, Haryana   

     floorNum facing       agePossession nearbyLocations  \
181       NaN    NaN  Under Construction             NaN   
2766      NaN    NaN  Under Construction             NaN   

                                                                                                                                                                                                                               description  \
181   A property by one of the most reputed builders. Construction given to l&t;. Ideal 6th floor , Three side open , Corner, Park/green facing aprtment. Simply the best location in the project. Current company rate is 6550 per sq ft.   
2766                                                                                                                                                                            The property is under construction it's by raheja revanta.   

     furnishDetails features  \
181              []      NaN   
2766             []      NaN   

                                                                                              rating  
181                                                                                              NaN  
2766  ['Safety4 out of 5', 'Lifestyle4 out of 5', 'Environment4 out of 5', 'Connectivity5 out of 5']

In [282]:
# replacing "Ground" with "0", "Lower Ground" with "0" and "Basement" with "-1"

df['floorNum'].str.split(' ').str.get(0).replace('Ground','0').str.replace('Lower', '0').str.replace('Basement', '-1').str.extract(r'(\d+)')

0
0       4
1       1
2      12
3       2
4       5
5       3
6       5
7      14
8       2
9      31
10      1
11      4
12      4
13      3
14     13
15     25
16      0
17      0
18     11
19     12
20      4
21     11
22      4
23      0
24     10
26      3
27     20
28      5
29      4
30      4
31      1
32     12
33      8
34     20
35      8
36     21
37      0
38      6
39      1
40      1
41     12
42      6
43      1
44      2
45      1
46      9
47      5
48     18
49     29
50      3
51      3
52      5
53      1
54      9
55      6
56      2
57     14
58      1
59      8
60      2
61     12
62      4
63      1
64      7
65      3
66      5
67      9
68     14
69     14
70     15
71      6
72      6
73      2
74      3
75      2
76      4
77      3
78      6
79      8
80      2
81      4
82      0
83     11
84      1
85     11
86     21
87      4
88      2
89      2
90      4
91     32
92      2
93      2
94      2
95     11
96     23
97      6
98      9
99      1
100     5
101    43
102     1
103     1
104     1
105    11
106    14
107     9
108     8
109     9
110     2
111     4
112    32
113    43
114     2
115     3
116     8
117     3
118    20
119     2
120     1
121     4
122     0
123     3
124     4
125     3
126     6
127    13
128    13
129    16
130     3
131     0
132    11
133     2
134    12
135     2
136     1
137     4
138    11
139     5
140     1
141    12
142    16
143     6
144     2
145     3
146    11
147    10
148     2
149     4
150    19
151     1
152     8
153    18
154    10
155     7
156     2
157    10
158     5
159    10
160    20
161    20
162     3
163    13
164     3
165     7
166     4
167     4
168     2
169     2
170     4
171     3
172     1
173     2
174    11
175    11
176    16
177     8
178     1
179    18
180     8
181   NaN
182     2
183    17
184    23
185    10
186    30
187     4
188     2
189     0
190     0
191     1
192     1
193    11
194    11
195     0
196    19
197    28
198     4
199     9
200     7
201     1
202    15
203     2
204     0
205     9
206     1
207     1
208     4
209     4
210     3
211     1
212     1
213     0
214    13
215     4
216     6
217    20
218    27
219     6
220    33
221    35
222    26
223     0
224     2
225     4
226    10
227    11
228    10
229     0
230     3
231     7
232     8
233     2
234     3
235     1
236     9
237    14
238     4
239     8
240    10
241     3
242     3
243     2
244     3
245     7
246     2
247     3
248     8
249    12
250     8
251     1
252     1
253    26
254    21
255     8
256     2
257     1
258     2
259    39
260     1
261     8
262     2
263     3
264     7
265     4
266    22
267    14
268    11
269    17
270     3
271    12
272     8
273     3
274    14
275     7
276     9
277     2
278     3
279     6
280    16
281    12
282     1
283     4
284    17
285     7
286    11
287     4
288     1
289     6
290     1
291    17
292     1
293    16
294    19
295    31
296    14
297     5
298     2
299     1
300     1
301    15
302     1
303     1
304     2
305    11
306     9
307    33
308    14
309     4
310     5
311    18
312     1
313    20
314     9
315    18
316     8
317     4
318     9
319     4
320     1
321     6
322     1
323    19
324    31
325     5
326     1
327     1
328    15
329     1
330     1
331     2
332    11
333     9
334    14
335     5
336     1
337     9
338    18
339     4
340     1
341    20
342    24
343     2
344     4
345    14
346     9
347    14
348     4
349    26
350     2
351    14
352    17
353     2
354    25
355     9
356     4
357     6
358     5
359     4
360    14
361    15
362     7
363     2
364    16
365    24
366    10
367     2
368     2
369     2
370     1
371     8
372    12
373    18
374     1
375     2
376     3
377    11
378     7
379     5
380     6
381    19
382    25
383    10
384    10
385     3
386     8
387     5
388     4
389     1
390    10
391    16
392    10
393     3
394    17
395     9
396    11
397    19
398     0
399     0
400     

In [283]:
df['floorNum'] = df['floorNum'].str.split(' ').str.get(0).replace('Ground','0').str.replace('Basement','-1').str.replace('Lower','0').str.extract(r'(\d+)')

In [284]:

df['floorNum'].value_counts()

floorNum
2     258
3     248
1     223
4     203
8     200
7     188
10    185
6     185
5     172
9     171
12    160
0     133
14    117
11    115
15     83
16     53
17     51
13     43
18     38
19     34
20     25
21     15
23     15
22     14
24     11
25      9
27      7
26      5
30      5
28      4
31      4
33      3
29      3
43      2
35      2
32      2
39      2
34      2
40      2
38      1
45      1
44      1
Name: count, dtype: int64

In [285]:
df.head()

property_name                 society  price  \
0     2 BHK Flat in Krishna Colony  maa bhagwati residency   0.45   
1        2 BHK Flat in Ashok Vihar            apna enclave   0.50   
2              2 BHK Flat in Sohna  tulsiani easy in homes   0.40   
3  2 BHK Flat in Sector 61 Gurgaon     smart world orchard   1.47   
4  2 BHK Flat in Sector 92 Gurgaon        parkwood westend   0.70   

   price_per_sqft                            areaWithType  bedRoom bathroom  \
0          5000.0          Carpet area: 900 (83.61 sq.m.)        2        2   
1          7692.0          Carpet area: 650 (60.39 sq.m.)        2        2   
2          6722.0          Carpet area: 595 (55.28 sq.m.)        2        2   
3         12250.0        Carpet area: 1200 (111.48 sq.m.)        2        2   
4          5204.0  Super Built up area 1345(124.95 sq.m.)        2        2   

  balcony additionalRoom                              address floorNum facing  \
0       1  not available     Krishna Colony, Gurgaon, Haryana        4   West   
1       1  not available   46b, Ashok Vihar, Gurgaon, Haryana        1   West   
2       3  not available              Sohna, Gurgaon, Haryana       12    NaN   
3       2     study room  Sector 61 Gurgaon, Gurgaon, Haryana        2    NaN   
4       3     study room  Sector 92 Gurgaon, Gurgaon, Haryana        5    NaN   

        agePossession  \
0     1 to 5 Year Old   
1        10+ Year Old   
2     0 to 1 Year Old   
3            Dec 2023   
4  Under Construction   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                nearbyLocations  \
0  ['Chintapurni Mandir', 'State bank ATM', 'Pearl Dental Clinic', 'Taneja Hospital', 'Dr. Ashok Jain', 'Dr. Madan Clinic', 'Geeta Nursing Home Gurgaon', 'Chiranjiv Hospital', 'Dev Man Kathuria Clinic', 'Aryan Hospital', 'Kr Dental Hub', 'Swastik Maternity and Medical Centre', 'Satyam Hospital Gurgaon', 'Ravi Clinic and Health Care Centre', 'D.R. Rajnis Gupta Clinic', 'Esic Hospital Gurugram', 'Gurgaon Eye Centre', 'Bindal Clinic', 'Shiv Mahima Patient Care Bureau', 'Shree Krishna Hospital Gurgaon', 'M.S Hospital', 'My Care Clinic', 'Ankur Clinic and Maternity Home', 'Triveni Hospital Gurgaon', 'Shri Gobind Hospital', 'Navjeevan Hospital and Maternity Centre', 'Lal Superspeciality Hospital', 'Sparsh Hospital Gurgaon', 'Prateek Nursing Home And Polyclinic', 'Jain Sant Phool Chand Ji Charitable Hospital', 'Sarvodya Hospital', 'Bhatnagar Maternity and Nursing Home', 'Tirath Ram Hospitals Pvt Ltd', 'Dr. Sarvejeet Singh', 'Centre For Sight Gurgaon New Railway Road', 'Dr. Ajay S. Gupta Clinic', 'Kidney Clinic', 'Mangalam Hospital and Heart Centre Gurgaon', 'Sethi Hospital Gurgaon', 'Shubham Hospital Gur

# Facing

In [286]:
df['facing']

0             West
1             West
2              NaN
3              NaN
4              NaN
5              NaN
6              NaN
7              NaN
8       North-East
9              NaN
10             NaN
11             NaN
12             NaN
13      South-East
14             NaN
15             NaN
16             NaN
17             NaN
18            East
19      North-East
20             NaN
21           North
22             NaN
23      South-West
24            East
26             NaN
27      North-East
28      South-East
29           North
30      North-East
31             NaN
32             NaN
33             NaN
34             NaN
35      North-East
36             NaN
37            East
38      North-East
39           North
40           North
41      North-West
42             NaN
43           North
44            East
45             NaN
46            East
47            West
48             NaN
49            East
50             NaN
51           North
52      North-East
53          

In [287]:
df['facing'].value_counts()

facing
North-East    505
East          490
North         301
South         203
West          183
North-West    162
South-East    144
South-West    135
Name: count, dtype: int64

In [288]:
df['facing'].isnull().sum()

874

- so found 874 rows have missing value in facing column.
- will replace missing value with "NA" 

In [289]:
df['facing'].fillna('NA', inplace=True)

In [290]:
df['facing'].value_counts()

facing
NA            874
North-East    505
East          490
North         301
South         203
West          183
North-West    162
South-East    144
South-West    135
Name: count, dtype: int64

# Address,AgePossession, Nearby Locations, Description, Furnish Details, Features , Rating 

- For these columns,more complex operations are needed, and they fall under the category of "Feature Engineering"
- Currently I am focusing on basic cleaning.
- However, I've implemented a small feature engineering step based on the "price" and "price_per_sqft"columns.
- I've introduced a new column, "Built-up Area of flat", calculated as follows:-
      Built-up Area of flat = (price * 1,00,00,000)/price_per_sqft
- This new column is placed at index location numer "4".
- This observation involves taking the price value, multiplying it by 1 crore (1,00,00,000), dissolving the year value, and then
dividing by the price per square foot.

In [291]:
df.insert(loc=4,column='area',value=round((df['price']*10000000)/df['price_per_sqft']))

In [292]:
df.head()

property_name                 society  price  \
0     2 BHK Flat in Krishna Colony  maa bhagwati residency   0.45   
1        2 BHK Flat in Ashok Vihar            apna enclave   0.50   
2              2 BHK Flat in Sohna  tulsiani easy in homes   0.40   
3  2 BHK Flat in Sector 61 Gurgaon     smart world orchard   1.47   
4  2 BHK Flat in Sector 92 Gurgaon        parkwood westend   0.70   

   price_per_sqft    area                            areaWithType  bedRoom  \
0          5000.0   900.0          Carpet area: 900 (83.61 sq.m.)        2   
1          7692.0   650.0          Carpet area: 650 (60.39 sq.m.)        2   
2          6722.0   595.0          Carpet area: 595 (55.28 sq.m.)        2   
3         12250.0  1200.0        Carpet area: 1200 (111.48 sq.m.)        2   
4          5204.0  1345.0  Super Built up area 1345(124.95 sq.m.)        2   

  bathroom balcony additionalRoom                              address  \
0        2       1  not available     Krishna Colony, Gurgaon, Haryana   
1        2       1  not available   46b, Ashok Vihar, Gurgaon, Haryana   
2        2       3  not available              Sohna, Gurgaon, Haryana   
3        2       2     study room  Sector 61 Gurgaon, Gurgaon, Haryana   
4        2       3     study room  Sector 92 Gurgaon, Gurgaon, Haryana   

  floorNum facing       agePossession  \
0        4   West     1 to 5 Year Old   
1        1   West        10+ Year Old   
2       12     NA     0 to 1 Year Old   
3        2     NA            Dec 2023   
4        5     NA  Under Construction   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                nearbyLocations  \
0  ['Chintapurni Mandir', 'State bank ATM', 'Pearl Dental Clinic', 'Taneja Hospital', 'Dr. Ashok Jain', 'Dr. Madan Clinic', 'Geeta Nursing Home Gurgaon', 'Chiranjiv Hospital', 'Dev Man Kathuria Clinic', 'Aryan Hospital', 'Kr Dental Hub', 'Swastik Maternity and Medical Centre', 'Satyam Hospital Gurgaon', 'Ravi Clinic and Health Care Centre', 'D.R. Rajnis Gupta Clinic', 'Esic Hospital Gurugram', 'Gurgaon Eye Centre', 'Bindal Clinic', 'Shiv Mahima Patient Care Bureau', 'Shree Krishna Hospital Gurgaon', 'M.S Hospital', 'My Care Clinic', 'Ankur Clinic and Maternity Home', 'Triveni Hospital Gurgaon', 'Shri Gobind Hospital', 'Navjeevan Hospital and Maternity Centre', 'Lal Superspeciality Hospital', 'Sparsh Hospital Gurgaon', 'Prateek Nursing Home And Polyclinic', 'Jain Sant Phool Chand Ji Charitable Hospital', 'Sarvodya Hospital', 'Bhatnagar Maternity and Nursing Home', 'Tirath Ram Hospitals Pvt Ltd', 'Dr. Sarvejeet Singh', 'Centre For Sight Gurgaon New Railway Road', 'Dr. Ajay S. Gupta Clinic', 'Kidney Clinic', 'Mangalam Hospital and Heart Centre Gurgaon',


- so now we have built up area value for every flat.

- will add another column by name "propert_type" on index value "1"


In [293]:
df.insert(loc=1,column='property_type',value='flat')

In [294]:
df.head()

property_name property_type                 society  \
0     2 BHK Flat in Krishna Colony          flat  maa bhagwati residency   
1        2 BHK Flat in Ashok Vihar          flat            apna enclave   
2              2 BHK Flat in Sohna          flat  tulsiani easy in homes   
3  2 BHK Flat in Sector 61 Gurgaon          flat     smart world orchard   
4  2 BHK Flat in Sector 92 Gurgaon          flat        parkwood westend   

   price  price_per_sqft    area                            areaWithType  \
0   0.45          5000.0   900.0          Carpet area: 900 (83.61 sq.m.)   
1   0.50          7692.0   650.0          Carpet area: 650 (60.39 sq.m.)   
2   0.40          6722.0   595.0          Carpet area: 595 (55.28 sq.m.)   
3   1.47         12250.0  1200.0        Carpet area: 1200 (111.48 sq.m.)   
4   0.70          5204.0  1345.0  Super Built up area 1345(124.95 sq.m.)   

   bedRoom bathroom balcony additionalRoom  \
0        2        2       1  not available   
1        2        2       1  not available   
2        2        2       3  not available   
3        2        2       2     study room   
4        2        2       3     study room   

                               address floorNum facing       agePossession  \
0     Krishna Colony, Gurgaon, Haryana        4   West     1 to 5 Year Old   
1   46b, Ashok Vihar, Gurgaon, Haryana        1   West        10+ Year Old   
2              Sohna, Gurgaon, Haryana       12     NA     0 to 1 Year Old   
3  Sector 61 Gurgaon, Gurgaon, Haryana        2     NA            Dec 2023   
4  Sector 92 Gurgaon, Gurgaon, Haryana        5     NA  Under Construction   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                nearbyLocations  \
0  ['Chintapurni Mandir', 'State bank ATM', 'Pearl Dental Clinic', 'Taneja Hospital', 'Dr. Ashok Jain', 'Dr. Madan Clinic', 'Geeta Nursing Home Gurgaon', 'Chiranjiv Hospital', 'Dev Man Kathuria Clinic', 'Aryan Hospital', 'Kr Dental Hub', 'Swastik Maternity and Medical Centre', 'Satyam Hospital Gurgaon', 'Ravi Clinic and Health Care Centre', 'D.R. Rajnis Gupta Clinic', 'Esic Hospital Gurugram', 'Gurgaon Eye Centre', 'Bindal Clinic', 'Shiv Mahima Patient Care Bureau', 'Shree Krishna Hospital Gurgaon', 'M.S Hospital', 'My Care Clinic', 'Ankur Clinic and Maternity Home', 'Triveni Hospital Gurgaon', 'Shri Gobind Hospital', 'Navjeevan Hospital and Maternity Centre', 'Lal Superspeciality Hospital', 'Sparsh Hospital Gurgaon', 'Prateek Nursing Home And Polyclinic', 'Jain Sant Phool Chand Ji Charitable Hospital', 'Sarvodya Hospital', 'Bhatnagar Maternity and Nursing Home', 'Tirath Ram Hospitals Pvt Ltd', 'Dr. Sarvejeet Singh', 'Centre For Sight Gurgaon New Railway Road', 'Dr. A

In [295]:
df.shape

(2997, 20)

- so now I have 2997 rows with 20 column.

- now again check current set of dataset by df.info


In [296]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2997 entries, 0 to 3016
Data columns (total 20 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   property_name    2997 non-null   object 
 1   property_type    2997 non-null   object 
 2   society          2997 non-null   object 
 3   price            2996 non-null   float64
 4   price_per_sqft   2996 non-null   float64
 5   area             2996 non-null   float64
 6   areaWithType     2997 non-null   object 
 7   bedRoom          2997 non-null   int64  
 8   bathroom         2997 non-null   string 
 9   balcony          2997 non-null   object 
 10  additionalRoom   2997 non-null   object 
 11  address          2991 non-null   object 
 12  floorNum         2995 non-null   object 
 13  facing           2997 non-null   object 
 14  agePossession    2996 non-null   object 
 15  nearbyLocations  2906 non-null   object 
 16  description      2997 non-null   object 
 17  furnishDetails   22

## Now will import this entire work as new dataset  name given "flats_cleaned" 

In [297]:
df.to_csv('flats_cleaned.csv',index=False)